src/scanner.rs

```diff
@@ -0,0 +1,564 @@
+    pub executable_path: PathBuf,
+
+    /// Path to the metadata source (embedded or sidecar file)
+    pub metadata_source: MetadataSource,
+}
+
+/// Represents the source of tool metadata.
+#[derive(Debug, Clone, PartialEq)]
+pub enum MetadataSource {
+    /// Metadata is embedded in the executable file itself
+    Embedded(PathBuf),
+
+    /// Metadata is in a sidecar `.yaml` file
+    Sidecar(PathBuf),
+}
+
+/// Errors that can occur during directory scanning.
+#[derive(Debug, thiserror::Error)]
+pub enum ScanError {
+    #[error("IO error scanning directory: {0}")]
+    IoError(#[from] std::io::Error),
+
+    #[error("Permission denied accessing path: {path}")]
+    PermissionDenied { path: PathBuf },
+}
+
+/// Scanner that traverses directories to discover potential tools.
+///
+/// The `DirectoryScanner` identifies executable files and their associated
+/// metadata sources (embedded or sidecar YAML files) while handling
+/// permission errors gracefully.
+///
+/// # Examples
+///
+/// ```
+/// use mcp_serve::scanner::DirectoryScanner;
+/// use std::path::Path;
+///
+/// let mut scanner = DirectoryScanner::new();
+/// // Use examples/tools if it exists, otherwise use current directory
+/// let tools_dir = if Path::new("examples/tools").exists() {
+///     Path::new("examples/tools")
+/// } else {
+///     Path::new(".")
+/// };
+/// let tools = scanner.scan_directory(tools_dir).unwrap();
+///
+/// for tool in tools {
+///     println!("Found tool: {}", tool.executable_path.display());
+/// }
+/// ```
+pub struct DirectoryScanner {
+    /// Collected errors during scanning
+    errors: Vec<ScanError>,
+}
+
+impl DirectoryScanner {
+    /// Create a new directory scanner.
+    pub fn new() -> Self {
+        Self { errors: Vec::new() }
+    }
+
+    /// Scan a directory for discoverable tools.
+    ///
+    /// Returns all discovered tools and collects any permission errors
+    /// encountered during scanning. Permission errors are stored internally
+    /// and can be retrieved with `take_errors()`.
+    ///
+    /// # Arguments
+    ///
+    /// * `directory` - The directory to scan for tools
+    ///
+    /// # Returns
+    ///
+    /// A vector of discovered tools, or an error if the directory
+    /// cannot be read or traversed.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use mcp_serve::scanner::DirectoryScanner;
+    /// use std::path::Path;
+    ///
+    /// let mut scanner = DirectoryScanner::new();
+    /// match scanner.scan_directory(Path::new("./tools")) {
+    ///     Ok(tools) => println!("Found {} tools", tools.len()),
+    ///     Err(e) => eprintln!("Scan failed: {}", e),
+    /// }
+    ///
+    /// // Check for permission errors that were collected
+    /// let errors = scanner.take_errors();
+    /// for error in errors {
+    ///     eprintln!("Warning: {}", error);
+    /// }
+    /// ```
+    pub fn scan_directory(&mut self, directory: &Path) -> Result<Vec<DiscoveredTool>, ScanError> {
```

Comment:
I don't think `Result` is the right type for this - we need to support cases where tools are discovered AND there are still errors

Reply:
Some errors are not recoverable (e.g. `IoError`), but others (like encountering a malformed tool script, missing metadata, etc) are